flink-core/src/main/java/org/apache/flink/core/classloading/ComponentClassLoader.java

```diff
@@ -0,0 +1,246 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.core.classloading;
+
+import org.apache.flink.util.function.FunctionWithException;
+
+import org.apache.flink.shaded.guava18.com.google.common.collect.Iterators;
+
+import java.io.IOException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.util.Arrays;
+import java.util.Enumeration;
+import java.util.Iterator;
+
+/**
+ * A {@link URLClassLoader} that restricts which classes can be loaded to those contained within the
+ * given classpath, except classes from a given set of packages that are either loaded parent or
+ * child-first.
+ */
+public class ComponentClassLoader extends URLClassLoader {
```

Comment:
I like the current abstraction of having a basic class loader and then the specialized classloaders. This would eventually give us the opportunity to change the behavior of the subclasses without breaking code.
Do you envision that this class is used directly? If not, do you want to guard against that (docs, visibility, or abstract)?

Reply:
I don't _think_ it will be used directly; technically there's nothing speaking against it because it is a fully functional CL after all.
The only downsides I would see (and also why I opted for subclasses instead of a static factory) is that it would make debugging CL leaks more difficult if all CLs were to just be ComponentClassLoaders. (Because you cant filter out plugin loaders for example).
So maybe it should indeed be an abstract class.